### PR TITLE
Helper methods for C++ Tooling

### PR DESCRIPTION
**Pull request by jmachowinski**
The C++ tooling relies on this helper method to
configure the transformer stack. In general it has 
nice properties and might also be useful for the
ruby tooling.

### CODE REVIEW

File: rock/orogen_plugin.rb [hunk 1]
```diff
@@ -727,6 +727,32 @@ def update_spec
...
+        base::samples::RigidBodyState curTr;
+        curTr.sourceFrame = (*transform)->getSourceFrame();
+        curTr.targetFrame = (*transform)->getTargetFrame();
+        
+        ret.push_back(curTr);
+    }
+    
+    return ret;"
+
+            task.hidden_operation("getNeededTransformations", transformationBody).
```
Comment (doudou) [inline rock/orogen_plugin.rb]: ??? I makes no sense to return a RBS if you only want to return a pair of source/target names. This is very confusing.

**Comment by doudou** [comment]
FYI, the Ruby tooling really has no need for this as it already knows everything _without having even the need to run anything_.

**Comment by jmachowinski** [comment]
> FYI, the Ruby tooling really has no need for this as it already knows everything without having even the need to run anything.

The ruby tooling would not know anything about transformations, that got registered inside of the Task code, without using the orogen plugin. Also this method allows to set up a transformer, 
without having the model files around.

File: rock/orogen_plugin.rb (hunk 1 above)
Comment (jmachowinski) [inline rock/orogen_plugin.rb]: Hm, 
I could also add an extra type for this...

**Comment by jmachowinski** [comment]
updated

**Comment by doudou** [comment]
Rock's core design and core proposition is to BE model based. If your goal is to break that, you'll have to pry Rock out of my cold, dead hands.

This information is static and should not be exposed by operations. Please use the orogen_metadata plugin to expose it, or (better) make orogen store some information in a format that you can load from C++ (i.e. YAML or JSON)

File: rock/orogen_plugin.rb (hunk 1 above)
Comment (goldhoorn) [inline rock/orogen_plugin.rb]: This is borken i just created a patch (_untested_) https://github.com/goldhoorn/drivers-orogen-transformer/commit/5f1e3725b4fad1f7899a4605d7d0c6704de98dc7  which fixes the hidden operation but it depends on https://github.com/orocos-toolchain/orogen/pull/40

Comment (shoppel) [inline rock/orogen_plugin.rb]: After applying the patch i get this for several modules:

===== Internal error =======
Please report the following error to the orogen developers
/home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/tasks.rb:246:in `register_for_generation': a hidden operation must have a body (OroGen::InternalError)
        from /home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/tasks.rb:517:in`each'
        from /home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/tasks.rb:517:in `generate'
        from /home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/project.rb:675:in`block in generate'
        from /home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/project.rb:675:in `each'
        from /home/stefan/Software/software_spacebot_test/tools/orogen/lib/orogen/gen/project.rb:675:in`generate'
        from /home/stefan/Software/software_spacebot_test/tools/orogen/bin/orogen:171:in `<main>'

Comment (shoppel) [inline rock/orogen_plugin.rb]: Without the patch it compiles but Bundles.initialize throws:
tools/orogen/lib/orogen/spec/task_context.rb:770:in `hidden_operation': undefined method`base_body=' for #OroGen::Spec::Operation:0x0000000360c8d8 (NoMethodError)

**Comment by goldhoorn** [comment]
Maybe i can jump in here. Both of you are right.
- Rock is model based, BUT the models are limited, specially regarding dynamic_port and the transformer
  - The modeler (Syskit) must guess how a model looks like under a defined configuration, examples:
    - The needed transforms depend on the naming of of the config
    - The inport/oitput ports on the JointDispatcher depending on the config

This yields into the problem, that the task model needs to be guessed (or remodeled a second time) inside the modelling layer. From my point of view it makes sense to 'ask' (even on the modeling layer) how a task model looks like under a specific configuration.
I would not call it 'running' the component, sure it needs to be launched/deployed and configured, but the component does not actually run from a component sense. Therefore janosch only added a way do (double-) check the model if moddeling is used.

If not a strong modeling is used it is okay to use this method as a helper.
Not all rock-users uses model-checkin tools like syskit, even most users simply use scrips without (strong) moddeling capabilities. We should keep rock open for both ways.

**Comment by doudou** [comment]
@goldhoorn: 

One can (and should) be able to model complete layers without having a single bit of C++ code on his machine. Remember model packs ? There you go. Otherwise, you do NOT have a model-based system, only a system that happens to have some models lying around. Finally, at the Syskit level, the ports-to-configuration mapping is done in reverse: one queries a dynamic service and the component must be configured accordingly. 
- dynamic ports: this does not apply here because the fact that some dynamic ports get or do not get created under certain configurations is already checked and reported when the component is spawned. One could move the concept of dynamic services to the orogen layer, which would make the modelling layer's job easier, but also make orogen a lot harder to use as a result. Dynamic ports are rare enough that I drew the line in favor of the orogen user.
- transformer: your argument does not hold for the transformer AS LONG AS orogen is the one generating the transformer C++ code. The model in orogen gives a 1:1 information since HE IS THE ONE GENERATING THE BEHAVIOUR and the models describe this behaviour faithfully. There is NO uncertainty about transformer configuration in the modelling layer.

If you allow the C++ code to change the behaviour described in orogen, then you are not modelled based anymore AT ALL. No guarantee about anything. Honestly, just use ROS (tf, ironically, can be handled with the current transformer handling by the way). Or do you plan on creating a C++-to-model generator ? Good luck with that ...

**Comment by jmachowinski** [comment]
This method does exactly the same thing as the huge and really complicated ruby code.
You INSTANCIATE the component, apply the configuration and the method returns
the needed transformations. The nice thing here is that we use the real code path and
not some code that is guessing, what might happen if I write this to that property.

A nice side effect, if you do this at runtime, and some developer did foo inside his component,
it will also work.

**Comment by goldhoorn** [comment]
I would like to go also for a other way solution. 
Adding this method means 'missusing' the task as model. It would work but would go in the wrong direction.

If we go to develop a c++ solution (analogous for orocos.rb) we should spend the work on a c++ model based representation.

This c++ model could then be used to configure the transformer. By doing so we not doubles the code path for task <-> model. The model could then be (theoretical) queried how it looks like under a specific configuration. But this get's more farther than this PR, more a conceptional direction which should not discussed here without having a larger road map and view.

**Comment by skasperski** [comment]
Seems to be outdated, as there is orocos_cpp now.

**Pull request closed without merging**